High Latency Telemetry Support

acfloria opened the pull request:
Improvements of the Iridiumsbd Driver and add a fallback to the high latency telemetry.

The changes in more details:
- Iridiumsbd Driver
   - Add parameter for session timeout, read interval, and stacking time
   - Add `VERBOSE_INFO` macro
   - Add sbd session timeout
   - Update heartbeat after successful SBD session and fill it in the telemetry_status message
   - Check test command to avoid being stuck in the test state
   - Make class variables private
   - Reorder function so that in the header and source file they have the same order
   - Fix updating the rssi at a constant rate when no sbd session is scheduled
   - Add variable timout to read_at function
- Commander Module
    - Add the logic for the high latency telemetry fallback
- Mavlink Module
    - Enable/disable transmitting from all mavlink instances of a certain type using a `vehicle_command` uorb message.

Switching Logic:
- Only activate the high latency telemetry if the vehicle is armed.
- If all low latency links are lost and a high latency link is present:
    -  Activate the high latency telemetry
- If the high latency is active and a low latency link is regained:
    - Deactivate the high latency telemetry
- If all low latency links are lost and a high latency link is not present:
    - Telemetry failsafe action
- If all low and high latency links are lost:
    - Telemetry failsafe action
- If a low latency link is regained after all low and high latency links were lost:
    - Deactivate the high latency telemetry

TSC21 commented:
@acfloria this seems very interesting. Question: based on the latest iteration that's ongoing on Mavlink form HIGH_LATENCY msg, would this be ready to actually establish a two-way communication instead of just having the FCU sending the data?

dagar commented:
Generally looks good. I'll review in detail and test on my side. What testing have you done with real hardware so far?

@TSC21 if you have everything it would be great to test this as well.

dagar commented on src/modules/commander/commander.cpp:
```diff
@@ -2838,8 +2879,49 @@ Commander::run()
+				mavlink_log_critical(&mavlink_log_pub, "ALL LOW LATENCY DATA LINKS LOST, ACTIVATING HIGH LATENCY LINK");
+				vehicle_command_s vehicle_cmd;
+				vehicle_cmd.command = vehicle_command_s::VEHICLE_CMD_MAVLINK_ENABLE_SENDING;
+				vehicle_cmd.param1 = 6;
```
magic numbers

dagar commented on src/modules/mavlink/mavlink_main.cpp:
```diff
@@ -2204,6 +2213,21 @@ Mavlink::task_main(int argc, char *argv[])
+			if (vehicle_cmd.command == vehicle_command_s::VEHICLE_CMD_MAVLINK_ENABLE_SENDING) {
+				if (_mode == (int)round(vehicle_cmd.param1)) {
+					if (_transmitting_enabled != (int)vehicle_cmd.param2) {
+						if ((int)vehicle_cmd.param2)
```
braces

dagar commented on src/modules/mavlink/mavlink_main.cpp:
```diff
@@ -2204,6 +2213,21 @@ Mavlink::task_main(int argc, char *argv[])
+						if ((int)vehicle_cmd.param2)
+							PX4_INFO("Enable transmitting with mavlink instance of type %s on device %s", mavlink_mode_str(_mode), _device_name);
+						else
+							PX4_INFO("Disable transmitting with mavlink instance of type %s on device %s", mavlink_mode_str(_mode), _device_name);
```
braces

dagar commented:
I'm wondering if we need to generalize the link loss handling for all link types, and possibly push more of it into the mavlink module.
https://github.com/PX4/Firmware/pull/7985

acfloria commented:
I tested the switching logic and testing the signal strength with the real hardware. I haven't sent any message over the link as I do not have a line rented yet. I will test transmitting data when the overall system is ready for it (Mavlink and QGC)

TSC21 commented:
@acfloria can you please rebase?

TSC21 commented on msg/vehicle_status.msg:
```diff
@@ -62,6 +62,7 @@ bool rc_signal_lost				# true if RC reception lost
 uint8 rc_input_mode				# set to 1 to disable the RC input, 2 to enable manual control to RC in mapping.
 
 bool data_link_lost				# datalink to GCS lost
+bool high_latency_data_link_active # all low latency datalinks to GCS lost
```
indentation

TSC21 commented on src/drivers/iridiumsbd/IridiumSBD.h:
```diff
@@ -85,54 +85,14 @@ const char *satcom_state_string[4] = {"STANDBY", "SIGNAL CHECK", "SBD SESSION",
 
-#define SATCOM_TX_BUF_LEN				50		// TX buffer size - maximum for a SBD MO message is 340, but billed per 50
-#define SATCOM_RX_MSG_BUF_LEN			300		// RX buffer size for MT messages
+#define SATCOM_TX_BUF_LEN				340		// TX buffer size - maximum for a SBD MO message is 340, but billed per 50
```
indentation

TSC21 commented on src/drivers/iridiumsbd/iridiumsbd_params.c:
```diff
@@ -8,4 +8,25 @@
+ * @max 300
+ * @group Iridium SBD
+ */
+PARAM_DEFINE_INT32(ISBD_SBD_TIMEOUT, 60);
```
Is this the time for "datalink" timeout on the high latency link?

TSC21 commented on src/modules/commander/commander_params.c:
```diff
@@ -121,6 +121,31 @@ PARAM_DEFINE_INT32(COM_DL_LOSS_T, 10);
+ * @min 60
+ * @max 3600
+ */
+PARAM_DEFINE_INT32(COM_HLDL_LOSS_T, 120);
```
I got my answer here.

TSC21 commented on src/modules/mavlink/mavlink_main.h:
```diff
@@ -476,6 +476,7 @@ class Mavlink
 
 private:
 	int			_instance_id;
+	bool		_transmitting_enabled;
```
indentation

TSC21 commented:
@acfloria what's the deal with QRT that lead you to add the conditions?

acfloria commented:
>It looks good to me. I can test this when I have my line rental activated. This should be synced also with the update on the HIGH_LATENCY Mavlink msg.

I will also do some tests and rent a line once the HIGH_LATENCY message is merged on mavlink.

>@acfloria what's the deal with QRT that lead you to add the conditions?

It is necessary, as for one build version mavlink is not used. I do not recall by heart which version.

TSC21 commented:
@acfloria were you able to test this in the mean time with the new msg?

TSC21 commented:
> @TSC21 I just added the support for it and tested it in SITL.

I should be able to test this on my end with real HW this week.

vayu-julie commented:
I tested this on a Pixhawk 3 Pro with a RockBlock Mk2 on a bench.  The telemetry switching works as outlined in the switching logic.

acfloria commented:
@dagar this PR has been extensively tested with real hardware. Can you review the code?

acfloria commented:
I moved the telemetry logic out of the main loop and redid the switching tests.

dagar commented:
Looking good. If you revert the nuttx_px4fmu-v2_default.cmake change and rebase on master will it fit?

I'm also wondering if you should move SimpleAnalyzer to a general location.

acfloria commented:
Rebased to master and reverted the cmake changes.

I would be ok to move the SimpleAnalyzer somewhere else but I don't know where an appropriate place would be. Apart from that I am not sure if it is really helpful for other modules.

dagar commented on src/modules/commander/CMakeLists.txt:
```diff
@@ -37,6 +37,9 @@ px4_add_module(
 		-Wno-sign-compare # TODO: fix all sign-compare
+	INCLUDES
+		# required because of nested includes in mavlink_main.h
+		${PX4_SOURCE_DIR}/mavlink/include/mavlink
```
Can we get rid of this?

acfloria replied:
Sure, I forgot to remove it.

dagar commented on src/modules/commander/Commander.hpp:
```diff
@@ -101,6 +93,32 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
+	 * Update the telemetry status and the corresponding status variables.
+	 * Perform system checks when new telemetry link connected.
+	 */
+	void poll_telemetry_status(bool checkAirspeed, bool *hotplug_timeout);
```
Passing airspeed to telemetry status poll is pretty odd. I understand why it was temporarily necessary though. I hope to get that under control in commander soon.

acfloria replied:
I agree, but currently it is not avoidable and fixing it should be done in another PR in my opinion.

dagar commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -3955,186 +3967,599 @@ class MavlinkStreamHighLatency : public MavlinkStream
+
+	void update_data()
+	{
+		const float update_rate = 1.0f / (_mavlink->get_main_loop_delay() * 1e-6f);
```
I'm not sure if you can rely on this being the actual execution rate.

acfloria replied:
True, I checked it and the actual update rate is much lower. I changed it now to using the filtered update rate computed from the time elapsed. I am using the filtered rate instead of the actual one as the rate is only used to determine when switching from averaging to the moving average filter and therefore using the actual rate would case switching between the methods too much.

dagar commented:
Overall this looks quite good, although I'm wondering a bit about the mavlink messages design longer term. 

@bkueng could you please review the mavlink changes in this PR?

bkueng commented on src/modules/commander/Commander.hpp:
```diff
@@ -122,6 +114,32 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
+			int32_t datalink_loss_timeout, int32_t datalink_regain_timeout, bool *status_changed);
+
+	// telemetry variables
+	int _telemetry_subs[ORB_MULTI_MAX_INSTANCES] = {};
```
can you create a separate struct for these? Then you need only one array and can use a constructor or member initializer to initialize the fields

acfloria replied:
Good point. I added that struct.

bkueng commented on src/modules/mavlink/mavlink_main.cpp:
```diff
@@ -988,6 +990,7 @@ Mavlink::send_bytes(const uint8_t *buf, unsigned packet_len)
 		return;
 	}
 
+
```
can you please avoid adding unnecessary newlines?

acfloria replied:
Sure, removed.

acfloria commented on src/drivers/telemetry/iridiumsbd/IridiumSBD.cpp:
```diff
@@ -390,374 +455,347 @@ void IridiumSBD::sbdsession_loop(void)
+		if (*buffer == MAVLINK_PACKAGE_START) {
+			if (SATCOM_TX_BUF_LEN - _tx_buf_write_idx - (*(buffer + 1) + 8) < 0) {
+				_tx_buf_write_idx = 0;
+				mavlink_log_critical(&_mavlink_log_pub, "Deleting full TX buffer before writing new message");
```
This is just to debug as in high latency telemetry the status texts are not transmitted.

acfloria replied:
Agreed. I am filling the tx_buf field of the telemetry_status correctly so it can be observed by that variable if a reset happens. However, I think for debugging it might be useful to have a status message of the driver where the status can be logged in more details. I will add that as a future PR after this one here is merged.

acfloria commented on src/drivers/telemetry/iridiumsbd/IridiumSBD.cpp:
```diff
@@ -390,374 +455,347 @@ void IridiumSBD::sbdsession_loop(void)
+
+	// check if there is enough space for the incoming mavlink message
+	if (buflen == 6) {
+		if (*buffer == MAVLINK_PACKAGE_START) {
```
I agree, that is why I added the check just after that one to catch anything which does not fit into that scheme.
However, in high latency mavlink 1 should be used because it assures that the HIGH_LATENCY2 message is 50 bytes. This is desired as for example rockblock bills in 50 byte increments.

acfloria replied:
I adopted it to the way the protocol_splitter is doing it.

Mavlink 1 is not enforced but encouraged as the design of the HIGH_LATENCY2 stream was to create a 50 byte status message. Using Mavlink 2 this goal cannot be met due to the larger header. I added a note in the header file. Once the full high latency telemetry together with QGC is working there should be a documentation about it for example in the dev guide where this information should also be written.

bkueng commented on src/modules/mavlink/mavlink_messages.h:
```diff
@@ -61,4 +63,56 @@ class StreamListItem
+
+private:
+	unsigned int _n = 0;
+	float _window = 60.0f;
```
Can you document the units? Also for things like `update_rate`

bkueng commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -3954,186 +3966,607 @@ class MavlinkStreamHighLatency : public MavlinkStream
+
+	void update_airspeed(float update_rate)
+	{
+		airspeed_s airspeed = {};
```
There's no need to initialize this (below too)

bkueng commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -3954,186 +3966,607 @@ class MavlinkStreamHighLatency : public MavlinkStream
+
+		update_tecs_status(_update_rate_filtered);
+
+		update_battery_status(_update_rate_filtered);
```
Since you're passing a class member `_update_rate_filtered`, you can avoid adding an argument and use `_update_rate_filtered` directly in `update_battery_status`

acfloria replied:
Done

bkueng commented on src/modules/mavlink/mavlink_stream.h:
```diff
@@ -103,6 +103,11 @@ class MavlinkStream
 #endif
 
+	/**
+	 * Function to collect/update data for the streams when update() is called.
```
Can you document this a bit more? Like saying that it's called at a high rate, independent from the configured stream rate.

bkueng commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -3929,22 +3941,22 @@ class MavlinkStreamMountOrientation : public MavlinkStream
 };
 
-class MavlinkStreamHighLatency : public MavlinkStream
+class MavlinkStreamHighLatency2 : public MavlinkStream
```
since this is so large, it's almost worth splitting it into a separate file.

bkueng commented on src/modules/mavlink/mavlink_stream.cpp:
```diff
@@ -114,11 +118,12 @@ MavlinkStream::update(const hrt_abstime t)
+		// not sending anything multiple messages in a short time are sent.
 		if (sent) {
-			_last_sent = (interval > 0) ? _last_sent + interval : t;
+			_last_sent = ((interval > 0) && ((int64_t)(1.5 * interval) > dt)) ? _last_sent + interval : t;
```
Can you use `1.5f` to avoid converting to double?

bkueng commented on src/modules/mavlink/mavlink_stream.cpp:
```diff
@@ -114,11 +118,12 @@ MavlinkStream::update(const hrt_abstime t)
+		// If the interval is non-zero and dt is smaller than 1.5 times the interval
+		// do not use the actual time but increment at a fixed rate, so that processing delays do not
+		// distort the average rate. The check of the maximum interval is done to ensure that after a long time
+		// not sending anything multiple messages in a short time are sent.
```
> The check of the maximum interval is done to ensure that after a long time not sending anything multiple messages in a short time are sent.

Should this not be more like: The check of the maximum interval is done to ensure that after a long time not sending anything, sending multiple messages in a short time is avoided.

acfloria commented on src/modules/mavlink/mavlink_messages.h:
```diff
@@ -61,4 +63,56 @@ class StreamListItem
+};
+
+template<typename Tin, typename Tout>
+void convert_limit_safe(Tin in, Tout *out)
```
Unfortunately not as this function is used with a packed struct and therefore won't compile using a reference.

acfloria commented on src/modules/mavlink/mavlink_messages.h:
```diff
@@ -61,4 +63,56 @@ class StreamListItem
+	};
+
+	SimpleAnalyzer(Mode mode, float window = 60.0f);
+	virtual ~SimpleAnalyzer();
```
Removed

acfloria commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -117,12 +123,11 @@ cm_uint16_from_m_float(float m)
 
-void get_mavlink_mode_state(struct vehicle_status_s *status, uint8_t *mavlink_state,
-			    uint8_t *mavlink_base_mode, uint32_t *mavlink_custom_mode)
+void get_mavlink_custom_mode(struct vehicle_status_s *status, uint8_t *mavlink_base_mode,
```
Made constant and renamed to `get_mavlink_navigation_mode`

acfloria commented on src/modules/mavlink/mavlink_main.cpp:
```diff
@@ -2186,6 +2200,60 @@ Mavlink::task_main(int argc, char *argv[])
+						_transmitting_enabled_commanded = true;
+						send_ack = true;
+
+					} else if ((vehicle_cmd.param1 <= 0.5f)) {
```
Sure, changed

acfloria commented on src/modules/mavlink/mavlink_main.cpp:
```diff
@@ -2186,6 +2200,60 @@ Mavlink::task_main(int argc, char *argv[])
+				bool send_ack = false;
+
+				if (_mode == MAVLINK_MODE_IRIDIUM) {
+					if ((vehicle_cmd.param1 > 0.5f)) {
```
Removed

acfloria commented on src/modules/mavlink/mavlink_main.cpp:
```diff
@@ -2201,7 +2236,11 @@ Mavlink::task_main(int argc, char *argv[])
 				msg.target_component = command_ack.target_component;
 				current_command_ack = command_ack.command;
 
+				// TODO: always transmit the acknowledge once it is only sent over the instance the command is received
```
I have an idea how to fix it, but as it potentially can break things and is not an trivial change I would prefer to do it in a separate PR.

bkueng replied:
Ok that's fine.

acfloria commented on src/modules/mavlink/mavlink_mission.cpp:
```diff
@@ -479,6 +479,11 @@ MavlinkMissionManager::send_mission_item_reached(uint16_t seq)
 MavlinkMissionManager::send(const hrt_abstime now)
 {
+	// do not send anything over high latency communication
+	if (_mavlink->get_mode() == Mavlink::MAVLINK_MODE_IRIDIUM) {
```
Ok, changed it.

acfloria commented:
@bkueng I implemented the changes you requested. Can you review the fixes again?

bkueng commented on src/drivers/telemetry/iridiumsbd/IridiumSBD.h:
```diff
@@ -167,114 +146,194 @@ class IridiumSBD : public device::CDev
+	char _test_command[32];
+	hrt_abstime _test_timer = 0;
+
+	uint8_t _rx_command_buf[SATCOM_RX_COMMAND_BUF_LEN] = {0};
```
I suggest you change it to `= {}`. `{0}` is confusing, as for example using `{1}` would lead to only the first element being 1 and all others 0.

acfloria replied:
Changed

bkueng commented on src/modules/mavlink/mavlink_high_latency2.h:
```diff
@@ -0,0 +1,201 @@
+#include "mavlink_simple_analyzer.h"
+#include "mavlink_stream.h"
+
+#ifndef MAVLINK_HIGH_LATENCY2_H_
```
I suggest you use `#pragma once` instead

acfloria replied:
Changed

bkueng commented:
@dagar circleci seems to be broken
